distributed/shuffle/tests/test_shuffle.py

```diff
@@ -121,10 +127,61 @@ async def test_bad_disk(c, s, a, b):
     # clean_scheduler(s)
 
 
-@pytest.mark.skip
+async def wait_until_worker_has_tasks(
+    prefix: str, worker: str, count: int, scheduler: Scheduler, interval: float = 0.01
+) -> None:
+    ws = scheduler.workers[worker]
+    while (
+        len(
+            [
+                key
+                for key, ts in scheduler.tasks.items()
+                if prefix in key and ts.state == "memory" and ws in ts.who_has
+            ]
+        )
+        < count
+    ):
+        await asyncio.sleep(interval)
+
+
+async def wait_for_tasks_in_state(
+    prefix: str,
+    state: str,
+    count: int,
+    dask_worker: Worker | Scheduler,
+    interval: float = 0.01,
+) -> None:
+    tasks: Mapping[str, SchedulerTaskState | WorkerTaskState]
+
+    if isinstance(dask_worker, Worker):
+        tasks = dask_worker.state.tasks
+    elif isinstance(dask_worker, Scheduler):
+        tasks = dask_worker.tasks
+    else:
+        raise TypeError(dask_worker)
+
+    while (
+        len([key for key, ts in tasks.items() if prefix in key and ts.state == state])
+        < count
+    ):
+        await asyncio.sleep(interval)
+
+
+class RemovedEventShuffleSchedulerExtension(ShuffleSchedulerExtension):
+    def __init__(self, scheduler: Scheduler):
+        super().__init__(scheduler)
+        self.removed_worker_event = asyncio.Event()
+
+    async def remove_worker(self, scheduler: Scheduler, worker: str) -> None:
+        await super().remove_worker(scheduler, worker)
+        self.removed_worker_event.set()
```

Comment:
This feels like a very specific hook we're listening for. IIUC this is only used to wait to assert the cleanup. However, why is the cleanup coupled to this hook specifically? I could see many reasons why something is only cleaned up _later_.
I think mocking is perfectly fine if the API we're mocking is stable. This doesn't feel stable and I'd rather not introduce this mock.

Thinking about the stable behavior we want to assert about the cleanup, this flow comes to mind

- Shuffle finishes/errs
- Shuffle eventually closes
- After shuffle is closed, there is no leftover state

I think this chain is more or less universal but there is no remove_worker in there. I suggest one of two things

1. If possible, expose the above control flow via an API and assert on that. For instance, our server classes all have stopped events / finished which can be awaited to signal that the shutdown is over. Is something similar possible here?
2. Alternatively, poll and see if the state is cleaned up (the typical `if not condition; await sleep(...)`)